clippy_utils/src/consts.rs

```diff
@@ -529,6 +530,59 @@ impl<'a, 'tcx> ConstEvalLateContext<'a, 'tcx> {
         }
     }
 
+    /// Simple constant folding to determine if an expression depends on an external crate
+    /// excluding crates listed in exceptions
+    pub fn expr_is_external(&mut self, e: &Expr<'_>, exceptions: &[Symbol]) -> bool {
```

Comment:
Let's add a new variant and check it in `expr`/`fetch_path_and_apply` rather than revisiting after the fact. It can be just specific to `core` as that's all we're using it for currently. `ConstantSource::CoreConstant` or similar